nexus/db-queries/src/db/datastore/inventory.rs

```diff
@@ -0,0 +1,914 @@
+            .select((
+                inv_collection2.field(db::schema::inv_collection::id),
+                diesel::dsl::count(
+                    db::schema::inv_collection_error::inv_collection_id
+                        .nullable(),
+                ),
+            ))
+            .order_by(
+                inv_collection2
+                    .field(db::schema::inv_collection::time_started)
+                    .asc(),
+            )
+            .load_async(&*conn)
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+            .internal_context("listing oldest collections")?;
+
+        if u32::try_from(candidates.len()).unwrap() <= nkeep {
+            debug!(
+                &opctx.log,
+                "inventory_prune_one: nothing eligible for removal (too few)";
+                "candidates" => ?candidates,
+            );
+            return Ok(None);
+        }
+
+        // We've now got up to "nkeep + 1" oldest collections, starting with the
+        // very oldest.  We can get rid of the oldest one unless it's the only
+        // complete one.  Another way to think about it: find the _last_
+        // complete one.  Remove it from the list of candidates.  Now mark the
+        // first item in the remaining list for deletion.
+        let last_completed_idx = candidates
+            .iter()
+            .enumerate()
+            .rev()
+            .find(|(_i, (_collection_id, nerrors))| *nerrors == 0);
+        let candidate = match last_completed_idx {
+            Some((i, _)) if i == 0 => candidates.iter().skip(1).next(),
+            _ => candidates.iter().next(),
+        }
+        .map(|(collection_id, _nerrors)| *collection_id);
+        if let Some(c) = candidate {
+            debug!(
+                &opctx.log,
+                "inventory_prune_one: eligible for removal";
+                "collection_id" => c.to_string(),
+                "candidates" => ?candidates,
+            );
+        } else {
+            debug!(
+                &opctx.log,
+                "inventory_prune_one: nothing eligible for removal";
+                "candidates" => ?candidates,
+            );
+        }
+        Ok(candidate)
+    }
+
+    /// Removes an inventory collection from the database
+    ///
+    /// The caller of this (non-pub) function is responsible for authz.
+    async fn inventory_delete_collection(
+        &self,
+        opctx: &OpContext,
+        collection_id: Uuid,
+    ) -> Result<(), Error> {
+        // As with inserting a whole collection, we remove it in one big
+        // transaction for simplicity.  Similar considerations apply.  We could
+        // break it up if these transactions become too big.  But we'd need a
+        // way to stop other clients from discovering a collection after we
+        // start removing it and we'd also need to make sure we didn't leak a
+        // collection if we crash while deleting it.
```

Comment:
Do we also need to prune no-longer-referenced `hw_baseboard_id` or `sw_caboose` rows? It's a little hard to imagine `hw_baseboard_id` getting "too big" since it only gets a row for each physical component the rack sees, but maybe in a large/long-lived multirack deployment? `sw_caboose` gets a few new rows for every update, so probably grows somewhat faster but still not all that quickly.

Reply:
In principle, eventually, yes.  I  think this is not urgent and we will notice before it becomes so.